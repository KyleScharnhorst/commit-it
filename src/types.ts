import Enquirer from 'enquirer'
import { PromptOptions } from './GitEmoji'

export interface CommitItOptions {
  dryRun?: boolean
  messageSeparator?: string
  // eslint-disable-next-line no-use-before-define
  plugins?: CommitItPlugin[]
  titleSeparator?: string
}

export interface EnquirerInterface {
  prompt(questions: PromptOptions): Promise<any>
}

export abstract class CommitItPlugin {
  public name: string
  public enquirer: EnquirerInterface
  protected constructor(name: string, enquirer?: EnquirerInterface) {
    this.name = name
    if (enquirer) {
      this.enquirer = enquirer
    } else {
      ;(this as any).enquirer = new Enquirer()
    }
  }

  public abstract load(
    options: CommitItOptions,
    commit: any
  ): Promise<[any, any]>

  public abstract run(
    options: CommitItOptions,
    commit: any
  ): Promise<[any, any]>
}

export interface Commit {
  body?: {
    plugin: string
    text: string
  }[]
  data?: Record<string, any>
  title?: {
    plugin: string
    text: string
  }[]
}
