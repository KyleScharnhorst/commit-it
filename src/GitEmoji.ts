import {
  Commit,
  CommitItOptions,
  CommitItPlugin,
  EnquirerInterface
} from './types'

export interface GitEmojiOptions {
  areas?: {
    description?: string
    text: string
  }[]
  areasRequired?: boolean
  askForShortDescription?: boolean
  commitBodyRequired?: boolean
  formatBody?: ({
    shortDescription,
    bodyText,
    commitBody,
    areas,
    intentions
  }: {
    areas: { choice: string; description?: string; text: string }[]
    bodyText: string
    commitBody: string
    intentions: {
      choice: string
      description?: string
      text: string
    }[]
    shortDescription: string
  }) => string
  formatTitle?: ({
    shortDescription,
    titleText,
    commitBody,
    areas,
    intentions
  }: {
    areas: { choice: string; description?: string; text: string }[]
    commitBody: string
    intentions: {
      choice: string
      description?: string
      text: string
    }[]
    shortDescription: string
    titleText: string
  }) => string
  intentions?: {
    description?: string
    text: string
  }[]
  multipleAreas?: boolean
  multipleIntentions?: boolean
  titleSeparator?: string
}

export const gitEmojiDefaultOptions: GitEmojiOptions = {
  titleSeparator: '',
  multipleIntentions: true,
  multipleAreas: true,
  commitBodyRequired: true,
  areasRequired: false,
  intentions: [
    {
      description: 'Initial commit',
      text: '🎉'
    },
    {
      description: 'Version tag',
      text: '🔖'
    },
    {
      description: 'New feature',
      text: '✨'
    },
    {
      description: 'Bugfix',
      text: '🐛'
    },
    {
      description: 'Metadata',
      text: '📇'
    },
    {
      description: 'Documentation',
      text: '📚'
    },
    {
      description: 'Documenting source code',
      text: '💡'
    },
    {
      description: 'Performance',
      text: '🐎'
    },
    {
      description: 'Cosmetic',
      text: '💄'
    },
    {
      description: 'Tests',
      text: '🚨'
    },
    {
      description: 'Adding a test',
      text: '✅'
    },
    {
      description: 'Make a test pass',
      text: '✔️'
    },
    {
      description: 'General update',
      text: '⚡'
    },
    {
      description: 'Improve format/structure',
      text: '🎨'
    },
    {
      description: 'Refactor code',
      text: '🔨'
    },
    {
      description: 'Removing code/files',
      text: '🔥'
    },
    {
      description: 'Continuous Integration',
      text: '💚'
    },
    {
      description: 'Security',
      text: '🔒'
    },
    {
      description: 'Upgrading dependencies',
      text: '⬆️'
    },
    {
      description: 'Downgrading dependencies',
      text: '⬇️'
    },
    {
      description: 'Lint',
      text: '👕'
    },
    {
      description: 'Translation',
      text: '👽'
    },
    {
      description: 'Text',
      text: '📝'
    },
    {
      description: 'Critical hotfix',
      text: '🚑'
    },
    {
      description: 'Deploying stuff',
      text: '🚀'
    },
    {
      description: 'Fixing on MacOS',
      text: '🍎'
    },
    {
      description: 'Fixing on Linux',
      text: '🐧'
    },
    {
      description: 'Fixing on Windows',
      text: '🏁'
    },
    {
      description: 'Work in progress',
      text: '🚧'
    },
    {
      description: 'Adding CI build system',
      text: '👷'
    },
    {
      description: 'Analytics or tracking code',
      text: '📈'
    },
    {
      description: 'Removing a dependency',
      text: '➖'
    },
    {
      description: 'Adding a dependency',
      text: '➕'
    },
    {
      description: 'Docker',
      text: '🐳'
    },
    {
      description: 'Configuration files',
      text: '🔧'
    },
    {
      description: 'Package.json in JS',
      text: '📦'
    },
    {
      description: 'Merging branches',
      text: '🔀'
    },
    {
      description: 'Bad code / need improv.',
      text: '💩'
    },
    {
      description: 'Reverting changes',
      text: '⏪'
    },
    {
      description: 'Breaking changes',
      text: '💥'
    },

    {
      description: 'Code review changes',
      text: '👌'
    },

    {
      description: 'Accessibility',
      text: '♿'
    },
    {
      description: 'Move/rename repository',
      text: '🚚'
    }
  ]
}

export interface PromptOptions {
  [key: string]: any
  name: string
}

export class GitEmoji extends CommitItPlugin {
  private readonly options: GitEmojiOptions
  constructor(options?: GitEmojiOptions, enquirer?: EnquirerInterface) {
    super('GitEmoji', enquirer)
    this.options = {
      ...gitEmojiDefaultOptions,
      ...options
    }
  }

  public async load(
    config: CommitItOptions,
    commit: Commit
  ): Promise<[any, any]> {
    commit.data.intentions = [
      ...(commit.data.intentions || []),
      ...(this.options.intentions || [])
    ].filter(Boolean)

    commit.data.areas = [
      ...(commit.data.areas || []),
      ...(this.options.areas || [])
    ].filter(Boolean)
    return [config, commit]
  }

  public async run(options, commit: Commit): Promise<[any, any]> {
    const allIntentions = commit.data.intentions
      .filter((x) => x?.text)
      .map((x) => ({
        ...x,
        choice: `${x.text || ''} ${x.description || ''}`.trim()
      }))

    const allAreas =
      commit.data.areas
        ?.filter((x) => x?.text)
        .map((x) => ({
          ...x,
          choice: `${x.text || ''} ${x.description || ''}`.trim()
        })) || []

    const intentions = (
      await this.enquirer.prompt({
        name: 'selectedIntention',
        type: 'autocomplete',
        message: 'Intention:',
        multiple: true,
        choices: allIntentions.map((x) => x.choice)
      })
    ).selectedIntention
      .map((x) => allIntentions.find((y) => y.choice === x))
      .sort((a, b) => a.description.localeCompare(b.description))

    const noIntentions = !intentions || intentions.length === 0

    if (noIntentions) {
      throw new Error('No intention selected')
    }

    let areas
    if (allAreas.length) {
      areas = (
        await this.enquirer.prompt({
          name: 'selectedArea',
          type: 'autocomplete',
          message: 'Area:',
          multiple: true,
          choices: allAreas.map((x) => x.choice)
        })
      ).selectedArea
        ?.map((x) => allAreas.find((y) => y.choice === x))
        .sort((a, b) => a.text.localeCompare(b.text))
    }

    const noAreasSelected = !areas || !areas?.length

    if (this.options.areasRequired && noAreasSelected) {
      throw new Error('At least one area is required')
    }

    let shortDescription

    if (
      (noAreasSelected && this.options.askForShortDescription !== false) ||
      this.options.askForShortDescription
    ) {
      shortDescription = (
        await this.enquirer.prompt({
          type: 'input',
          name: 'shortDescription',
          message: 'Short description:'
        })
      ).shortDescription

      if (shortDescription) {
        areas = [{ text: shortDescription }]
      }
    }

    const {
      commitBody
    }: {
      commitBody: string
    } = await this.enquirer.prompt({
      type: 'input',
      name: 'commitBody',
      message: 'Message:'
    })

    if (this.options.commitBodyRequired && !commitBody?.trim()) {
      throw new Error('Commit body is required')
    }

    const titleText =
      intentions.map((x) => x.text).join(' ') +
      ' ' +
      (areas
        ? `${this.options.titleSeparator} ${areas
            .map((x) => x.text)
            .join(', ')}`.trim()
        : `${this.options.titleSeparator} ${commitBody}`
      ).trim()

    const formattedTitleText = this.options.formatTitle
      ? this.options.formatTitle({
          shortDescription,
          titleText,
          commitBody,
          areas,
          intentions
        })
      : titleText

    commit.title = [
      ...(commit.title || []),
      { plugin: this.name, text: formattedTitleText }
    ]

    const bodyText = [
      commitBody || '',
      intentions?.length > 0 && '**Intentions:**',
      intentions?.map((x) => `- ${x.choice}`).join('\n') || '',
      areas?.length > 0 && '**Areas:**',
      areas?.map((x) => `- ${x.text}`).join('\n') || ''
    ]
      .filter(Boolean)
      .join('\n\n')
      .trim()

    const formattedBodyText = this.options.formatBody
      ? this.options.formatBody({
          shortDescription,
          bodyText,
          commitBody,
          areas,
          intentions
        })
      : bodyText

    commit.body = [
      ...(commit.body || []),
      {
        plugin: this.name,
        text: formattedBodyText
      }
    ]

    return [options, commit]
  }
}
