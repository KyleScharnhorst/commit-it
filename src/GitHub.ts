import { Commit, CommitItOptions, EnquirerInterface } from './types'
import {
  IntentionAndAreaPlugin,
  IntentionAndAreaPluginOptions
} from './IntentionAndAreaPlugin'

export const githubDefaultOptions: IntentionAndAreaPluginOptions = {
  pluginId: 'github'
}

export class GitHub extends IntentionAndAreaPlugin {
  constructor(
    options?: Partial<IntentionAndAreaPluginOptions>,
    enquirer?: EnquirerInterface
  ) {
    super(
      {
        ...githubDefaultOptions,
        ...options
      },
      enquirer
    )
  }
}
