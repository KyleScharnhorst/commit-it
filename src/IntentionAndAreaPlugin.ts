import {
  Commit,
  CommitItOptions,
  CommitItPlugin,
  EnquirerInterface,
  PluginOptions
} from './types'

export interface IntentionAndAreaPluginOptions extends PluginOptions {
  areas?: {
    description?: string
    text: string
  }[]
  areasRequired?: boolean
  askForShortDescription?: boolean
  commitBodyRequired?: boolean
  formatBody?: ({
    shortDescription,
    bodyText,
    commitBody,
    areas,
    intentions
  }: {
    areas: { choice: string; description?: string; text: string }[]
    bodyText: string
    commitBody: string
    intentions: {
      choice: string
      description?: string
      text: string
    }[]
    shortDescription: string
  }) => string
  formatTitle?: ({
    shortDescription,
    titleText,
    commitBody,
    areas,
    intentions
  }: {
    areas: { choice: string; description?: string; text: string }[]
    commitBody: string
    intentions: {
      choice: string
      description?: string
      text: string
    }[]
    shortDescription: string
    titleText: string
  }) => string
  intentions?: {
    description?: string
    text: string
  }[]
  multipleAreas?: boolean
  multipleIntentions?: boolean
  sortAreas?: (a: { description; text }, b: { description; text }) => number
  sortIntentions?: (
    a: { description; text },
    b: { description; text }
  ) => number
  titleSeparator?: string
}

export class IntentionAndAreaPlugin extends CommitItPlugin {
  private readonly options: IntentionAndAreaPluginOptions
  constructor(
    options?: IntentionAndAreaPluginOptions,
    enquirer?: EnquirerInterface
  ) {
    super(
      {
        pluginId: 'IntentionAndAreaPlugin',
        ...options
      },
      enquirer
    )
    this.options = {
      ...options
    }
  }

  public async load(
    config: CommitItOptions,
    commit: Commit
  ): Promise<[any, any]> {
    commit.data.intentions = [
      ...(commit.data.intentions || []),
      ...(this.options.intentions || [])
    ].filter(Boolean)

    commit.data.areas = [
      ...(commit.data.areas || []),
      ...(this.options.areas || [])
    ].filter(Boolean)
    return [config, commit]
  }

  public async run(options, commit: Commit): Promise<[any, any]> {
    let allIntentions = commit.data.intentions
      .filter((x) => x?.text)
      .map((x) => ({
        ...x,
        choice: `${x.text || ''} ${x.description || ''}`.trim()
      }))

    if (this.options.sortIntentions) {
      allIntentions = allIntentions.sort(this.options.sortIntentions)
    }

    let allAreas =
      commit.data.areas
        ?.filter((x) => x?.text)
        .map((x) => ({
          ...x,
          choice: `${x.text || ''} ${x.description || ''}`.trim()
        })) || []

    if (this.options.sortAreas) {
      allAreas = allAreas.sort(this.options.sortAreas)
    }

    const intentions = (
      await this.enquirer.prompt({
        name: 'selectedIntention',
        type: 'autocomplete',
        message: 'Intention:',
        multiple: true,
        choices: allIntentions.map((x) => x.choice)
      })
    ).selectedIntention
      .map((x) => allIntentions.find((y) => y.choice === x))
      .sort((a, b) => a.description.localeCompare(b.description))

    const noIntentions = !intentions || intentions.length === 0

    if (noIntentions) {
      throw new Error('No intention selected')
    }

    let areas
    if (allAreas.length) {
      areas = (
        await this.enquirer.prompt({
          name: 'selectedArea',
          type: 'autocomplete',
          message: 'Area:',
          multiple: true,
          choices: allAreas.map((x) => x.choice)
        })
      ).selectedArea
        ?.map((x) => allAreas.find((y) => y.choice === x))
        .sort((a, b) => a.text.localeCompare(b.text))
    }

    const noAreasSelected = !areas || !areas?.length

    if (this.options.areasRequired && noAreasSelected) {
      throw new Error('At least one area is required')
    }

    let shortDescription

    if (
      (noAreasSelected && this.options.askForShortDescription !== false) ||
      this.options.askForShortDescription
    ) {
      shortDescription = (
        await this.enquirer.prompt({
          type: 'input',
          name: 'shortDescription',
          message: 'Short description:'
        })
      )?.shortDescription

      if (shortDescription) {
        areas = [{ text: shortDescription }]
      }
    }

    const {
      commitBody
    }: {
      commitBody: string
    } = await this.enquirer.prompt({
      type: 'input',
      name: 'commitBody',
      message: 'Message:'
    })

    if (this.options.commitBodyRequired && !commitBody?.trim()) {
      throw new Error('Commit body is required')
    }

    const titleText =
      intentions.map((x) => x.text).join(' ') +
      ' ' +
      (areas
        ? `${this.options.titleSeparator} ${areas
            .map((x) => x.text)
            .join(', ')}`.trim()
        : `${this.options.titleSeparator} ${commitBody}`
      ).trim()

    const formattedTitleText = this.options.formatTitle
      ? this.options.formatTitle({
          shortDescription,
          titleText,
          commitBody,
          areas,
          intentions
        })
      : titleText

    commit.title = [
      ...(commit.title || []),
      {
        pluginId: this.pluginId,
        pluginAction: this.pluginAction,
        text: formattedTitleText
      }
    ]

    const bodyText = [
      commitBody || '',
      intentions?.length > 0 && '**Intentions:**',
      intentions?.map((x) => `- ${x.choice}`).join('\n') || '',
      areas?.length > 0 && '**Areas:**',
      areas?.map((x) => `- ${x.text}`).join('\n') || ''
    ]
      .filter(Boolean)
      .join('\n\n')
      .trim()

    const formattedBodyText = this.options.formatBody
      ? this.options.formatBody({
          shortDescription,
          bodyText,
          commitBody,
          areas,
          intentions
        })
      : bodyText

    commit.body = [
      ...(commit.body || []),
      {
        pluginId: this.pluginId,
        pluginAction: this.pluginAction,
        text: formattedBodyText
      }
    ]

    return [options, commit]
  }
}
