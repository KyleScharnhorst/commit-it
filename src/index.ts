export * from './CommitIt'
export * from './GitEmoji'
export * from './GitHub'
export * from './IntentionAndAreaPlugin'
export * from './types'
