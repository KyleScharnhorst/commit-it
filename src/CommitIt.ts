import path from 'path'
import * as fs from 'fs'
import { GitEmoji } from './GitEmoji'
import simpleGit from 'simple-git'
import { Commit, CommitItOptions, CommitItPlugin } from './types'

export const DEFAULT_CONFIG_PATH = './commit.config.js'

export class CommitIt {
  private readonly options: CommitItOptions

  constructor(options: CommitItOptions = {}) {
    this.options = {
      messageSeparator: '\n',
      titleSeparator: '|',
      ...options
    }
  }

  public static async loadConfig(
    configPath: string = DEFAULT_CONFIG_PATH
  ): Promise<CommitItOptions> {
    const p = path.resolve(configPath)
    if (!fs.existsSync(p)) {
      return {
        plugins: [new GitEmoji()]
      }
    } else {
      const importedConfig = await import(p)

      return importedConfig.default
    }
  }

  public async initialize(): Promise<void> {}

  public async executePlugins(): Promise<string> {
    const plugins: CommitItPlugin[] = this.options.plugins || [new GitEmoji()]
    let options = this.options || {}
    let commit: Commit = {
      data: []
    }

    for (const plugin of plugins) {
      let result
      if (plugin.options.pluginAction !== 'run') {
        result = await plugin.load(options, commit)
        options = result[0]
        commit = result[1]
      }
    }

    for (const plugin of plugins) {
      let result
      if (plugin.options.pluginAction !== 'load') {
        result = await plugin.run(options, commit)
      }
      options = result[0]
      commit = result[1]
    }

    const title = commit.title
      ?.map((x) => x.text)
      .join(this.options.titleSeparator)
    const body = commit.body
      ?.map((x) => x.text)
      .join(this.options.messageSeparator)

    const message = (title + '\n\n' + body).trim()

    if (!message) {
      throw new Error('No commit message provided. Commit aborted.')
    }

    if (this.options.dryRun !== true) {
      await simpleGit()
        .outputHandler((cmd, stdout, stderr, ...args) => {
          const out = []
          stdout.on('data', (buffer) => out.push(buffer))
          stdout.on('close', () =>
            console.log(Buffer.concat(out).toString('utf8'))
          )
        })
        .commit(message)
    }

    return message
  }
}
