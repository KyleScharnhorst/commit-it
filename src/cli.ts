#!/usr/bin/env node

import { Command } from 'commander'
import * as path from 'path'
import { CommitIt, DEFAULT_CONFIG_PATH } from './CommitIt'

interface CommanderOptions {
  config?: string
  dryRun?: boolean
}

;(async () => {
  const program = new Command()

  program
    .option(
      '-c, --config <config>',
      'path to a js config file. Defaults to ./commit.config.js'
    )
    .option('-d, --dry-run', 'does not make a commit')

  const options: CommanderOptions = program.opts()

  const configPath = path.resolve(options.config || DEFAULT_CONFIG_PATH)

  const config = await CommitIt.loadConfig(configPath)

  const am = new CommitIt(config)

  am.initialize().then(async () => {
    const message = await am.executePlugins()
    console.log('\n' + message + '\n')
    process.exit(0)
  })
})()
