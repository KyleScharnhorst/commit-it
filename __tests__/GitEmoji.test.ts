import { CommitIt, GitEmoji } from '../src'

describe('GitEmoji', () => {
  it('commit with intentions, areas, and a body', async () => {
    const commitIt = new CommitIt({
      dryRun: true,
      plugins: [
        new GitEmoji(
          {
            areas: [
              {
                name: 'User settings'
              },
              {
                name: 'Dashboard'
              }
            ]
          },
          {
            async prompt(questions) {
              if (questions.name === 'selectedIntention') {
                return Promise.resolve({
                  selectedIntention: [
                    '✔️ - Make a test pass',
                    '✅ - Adding a test'
                  ]
                })
              }
              if (questions.name === 'selectedArea') {
                return Promise.resolve({
                  selectedArea: ['User settings', 'Dashboard']
                })
              }
              if (questions.name === 'commitBody') {
                return Promise.resolve({
                  commitBody: 'Testing for certain things'
                })
              }
            }
          }
        )
      ]
    })

    expect(await commitIt.execute()).toMatchSnapshot()
  })
  it('commit body required', async () => {
    const commitIt = new CommitIt({
      dryRun: true,
      plugins: [
        new GitEmoji(
          {
            commitBodyRequired: true
          },
          {
            async prompt(questions) {
              if (questions.name === 'selectedIntention') {
                return Promise.resolve({
                  selectedIntention: [
                    '✔️ - Make a test pass',
                    '✅ - Adding a test'
                  ]
                })
              }
              if (questions.name === 'commitBody') {
                return Promise.resolve({
                  commitBody: ''
                })
              }
              if (questions.name === 'shortDescription') {
                return Promise.resolve({
                  shortDescription: 'short description of commit'
                })
              }
            }
          }
        )
      ]
    })

    await expect(commitIt.execute()).rejects.toMatchSnapshot()
  })
  it('areas required', async () => {
    const commitIt = new CommitIt({
      dryRun: true,
      plugins: [
        new GitEmoji(
          {
            areasRequired: true
          },
          {
            async prompt(questions) {
              if (questions.name === 'selectedIntention') {
                return Promise.resolve({
                  selectedIntention: [
                    '✔️ - Make a test pass',
                    '✅ - Adding a test'
                  ]
                })
              }
              if (questions.name === 'selectedArea') {
                return Promise.resolve({
                  selectedArea: []
                })
              }
              if (questions.name === 'commitBody') {
                return Promise.resolve({
                  commitBody: 'Testing for certain things'
                })
              }
            }
          }
        )
      ]
    })

    await expect(commitIt.execute()).rejects.toMatchSnapshot()
  })

  it('custom title and body formatters', async () => {
    const commitIt = new CommitIt({
      dryRun: true,
      plugins: [
        new GitEmoji(
          {
            formatTitle({ intentions }) {
              return intentions.map((x) => x.choice).join(', ')
            },
            formatBody() {
              return 'Custom body'
            },
            areas: [
              {
                name: 'User settings'
              },
              {
                name: 'Dashboard'
              }
            ]
          },
          {
            async prompt(questions) {
              if (questions.name === 'selectedIntention') {
                return Promise.resolve({
                  selectedIntention: [
                    '✔️ - Make a test pass',
                    '✅ - Adding a test'
                  ]
                })
              }
              if (questions.name === 'selectedArea') {
                return Promise.resolve({
                  selectedArea: ['User settings', 'Dashboard']
                })
              }
              if (questions.name === 'commitBody') {
                return Promise.resolve({
                  commitBody: 'Testing for certain things'
                })
              }
            }
          }
        )
      ]
    })

    expect(await commitIt.execute()).toMatchSnapshot()
  })
  it('asks for a short description', async () => {
    const commitIt = new CommitIt({
      dryRun: true,
      plugins: [
        new GitEmoji(
          {
            commitBodyRequired: false,
            askForShortDescription: true
          },
          {
            async prompt(questions) {
              if (questions.name === 'selectedIntention') {
                return Promise.resolve({
                  selectedIntention: [
                    '✔️ - Make a test pass',
                    '✅ - Adding a test'
                  ]
                })
              }
              if (questions.name === 'commitBody') {
                return Promise.resolve({
                  commitBody: 'commit body goes here'
                })
              }
              if (questions.name === 'shortDescription') {
                return Promise.resolve({
                  shortDescription: 'short description of commit'
                })
              }
            }
          }
        )
      ]
    })

    await expect(commitIt.execute()).resolves.toMatchSnapshot()
  })
  it('askForShortDescription: false and commitBodyRequired: false', async () => {
    const commitIt = new CommitIt({
      dryRun: true,
      plugins: [
        new GitEmoji(
          {
            askForShortDescription: false,
            commitBodyRequired: false
          },
          {
            async prompt(questions) {
              if (questions.name === 'selectedIntention') {
                return Promise.resolve({
                  selectedIntention: [
                    '✔️ - Make a test pass',
                    '✅ - Adding a test'
                  ]
                })
              }
              if (questions.name === 'commitBody') {
                return Promise.resolve({
                  commitBody: 'commit body goes here'
                })
              }
            }
          }
        )
      ]
    })
    await expect(commitIt.execute()).resolves.toMatchSnapshot()
  })
})
