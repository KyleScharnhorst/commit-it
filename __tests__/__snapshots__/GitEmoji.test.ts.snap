// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitEmoji areas required 1`] = `[Error: At least one area is required]`;

exports[`GitEmoji askForShortDescription: false and commitBodyRequired: false 1`] = `
"✅ ✔️ | commit body goes here

commit body goes here

✅ Adding a test, ✔️ Make a test pass"
`;

exports[`GitEmoji asks for a short description 1`] = `
"✅ ✔️ | short description of commit

commit body goes here

✅ Adding a test, ✔️ Make a test pass

- short description of commit"
`;

exports[`GitEmoji commit body required 1`] = `[Error: Commit body is required]`;

exports[`GitEmoji commit with intentions, areas, and a body 1`] = `
"✅ ✔️ | Dashboard, User settings

Testing for certain things

✅ Adding a test, ✔️ Make a test pass

- Dashboard
- User settings"
`;

exports[`GitEmoji custom title and body formatters 1`] = `
"✅ Adding a test, ✔️ Make a test pass

Custom body"
`;
